### PR TITLE
perf(wasm-builder): invert path remapping logic

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
The ecosystem team can use the "production" profile to remove the username from the wasm binary.

### CODE REVIEW

File: utils/wasm-builder/src/lib.rs [hunk 1]
```diff
@@ -167,6 +177,25 @@ impl WasmBuilder {
         Ok(matched_features)
     }
 
+    fn need_remap_path(&self, enabled_features: &[String]) -> bool {
```
Comment (DennisInSky) [inline utils/wasm-builder/src/lib.rs:180]:
```suggestion
    fn need_path_remap(&self, enabled_features: &[String]) -> bool {
```

File: utils/wasm-builder/src/lib.rs [hunk 2]
```diff
@@ -167,6 +177,25 @@ impl WasmBuilder {
...
+            .iter()
+            .filter_map(|feature| crate_info.features.get(feature))
+            .flatten()
+            .any(|feature| feature == "gstd/debug");
+
+        let debug_in_dependencies = crate_info.dependencies.iter().any(|dependency| {
+            dependency.kind == DependencyKind::Normal
+                && dependency.name == "gstd"
+                && dependency.features.contains(&String::from("debug"))
+        });
```
Comment (DennisInSky) [inline utils/wasm-builder/src/lib.rs:194]: Could it be the `debug` feature is enabled via a transitive dependency like
`dapp` crate depends on some `dapp-lib` crate which in turn depends on `gstd` with the `debug` feature on

File: utils/wasm-builder/src/wasm_project.rs [hunk 3]
```diff
@@ -115,6 +115,11 @@ impl WasmProject {
         }
     }
 
+    /// Return the path to the temporary generated `Cargo.toml`.
```
Comment (DennisInSky) [inline utils/wasm-builder/src/wasm_project.rs:118]: The comment has been copy-pasted from the below method

File: utils/wasm-builder/src/lib.rs [hunk 4]
```diff
@@ -167,6 +177,25 @@ impl WasmBuilder {
...
     }
 
+    fn need_path_remap(&self, enabled_features: &[String]) -> bool {
+        let crate_info = CrateInfo::from_manifest(&self.wasm_project.original_manifest_path())
+            .expect("Infallible");
+
+        let debug_in_features = enabled_features
+            .iter()
+            .filter_map(|feature| crate_info.features.get(feature))
+            .flatten()
```
Comment (DennisInSky) [inline utils/wasm-builder/src/lib.rs:187]: Btw, there is the same question about transitive features

**Review by breathx — CHANGES_REQUESTED** [review]
Apply fixes for @DennisInSky comments

**Comment by StackOverflowExcept1on** [comment]
@breathx #2915 will add a panic handler feature to gstd. It makes sense to enable path remapping only when the panic handler is enabled.

If the user uses dependency (such as dapps-gear-lib) that references gstd, the user should disable the panic handler. Otherwise, such dependency will define the panic handler twice, and we will consider it semantically incorrect (also it leads to compile error).

Does it make sense to check transitive dependencies for the `gstd/debug` function in this case?

File: utils/wasm-builder/src/lib.rs [hunk 5]
```diff
@@ -107,7 +107,7 @@ impl WasmBuilder {
         let profile = if profile == "debug" { "dev" } else { profile };
         self.cargo.set_profile(profile.to_string());
         self.cargo.set_features(&self.enabled_features()?);
-        if env::var("__GEAR_WASM_BUILDER_NO_PATH_REMAP").is_err() && profile != "dev" {
+        if env::var("__GEAR_WASM_BUILDER_NO_PATH_REMAP").is_err() && profile == "production" {
```
Comment (DennisInSky) [inline utils/wasm-builder/src/lib.rs:110]: This more resembles exlicit feature and has less flexibility then a dedicated feature. If it is the only option we can suggest to our customers, I would suggest introducing the latter, so they can decide themselves whether and when they want the paths to be remapped.

**Review by breathx — APPROVED** [review]

**Review by DennisInSky — APPROVED** [review]

**Pull request merged**